0017-negotiation-and-execution-protocol.adoc

```diff
@@ -0,0 +1,544 @@
+This means, that some negotiation protocols could be re-used for multiple execution protocols if the execution protocols have the same input parameters.
+
+=== Invitation phase
+
+We introduce an invitation phase in order to enable decentralization.
+Makers don't necessarily have a central order book, the invitation helps them to overcome the problem of decentral offers getting stale.
+
+On an abstract level an invitation protocol can be defined as a function:
+
+```
+fn invitation() -> negotiation_params[] {
+    // steps (messages) needed to come from nothing to negotiation_params
+}
+```
+
+==== Communication
+
+Message Overview:
+
+![Invitation](http://www.plantuml.com/plantuml/proxy?src=https://raw.githubusercontent.com/comit-network/spikes/master/assets/0017-SequenceDiagram-InvitationPhase.puml&fmt=svg)
+
+==== Trade Invitation ====
+
+The trade invitation specifies which assets on which ledgers a maker _is willing_ to trade.
+The trade invitation is not necessarily part of the negotiation.
+Its purpose is to connect maker and taker.
+
+An asset-availability is defined by:
+
+[%header,cols=1*]
+|===
+|required
+
+|connection-info
+|bid-asset
+|ask-asset
+|===
+
+[%header,cols=1*]
+|===
+|optional
+
+|timestamp
+|bid-amount
+|ask-amount
+|===
+
+Constraints:
+
+* Connection-info currently consists of IP-address and libp2p peer-id
+* The optional parameters help takers to understand what a maker was capable of at a certain point in time.
+
+Representation in JSON:
+
+```json
+{
+    "connection-info": { ... },
+    "bid": {
+        "ledger": ...
+        "asset": ...
+    },
+    "ask": {
+        "ledger": ...
+        "asset": ...
+    }
+}
+```
+When designing the negotiation protocol
+One should aim for a standardized format for trade invitations to be able to process them in the negotiation phase.
+
+
+=== Negotiation protocol
+
+On an abstract level the negotiation protocol can be defined as a function:
+
+```
+fn negotiation(input_params[]) -> execution_params[] {
+    // steps (messages) needed to come from input_params to execution_params
+}
+```
+
+The output of the negotiation protocol should enable the execution phase.
+As we only have one execution protocol (RFC003 - Atomic Swaps using HTLCs) at the moment this negotiation protocol will be designed to output the parameters for that protocol.
+
+Asset-availability handling (i.e. building a central order-book) is not part of this protocol.
+
+The invitation phase may be part of an implementation of a negotiation protocol.
+
+==== Negotiation Input Parameters
+
+The minimum input parameters for the protocol are:
+
+* ledger/asset pair
+* connection-info (of the negotiation protocol)
+
+Example of one input parameter as JSON object:
+```
+{
+    "bid_ledger": ... ,
+    "bid_asset": ... ,
+    "ask_ledger": ... ,
+    "ask_asset": ... ,
+    "peer_id" ...
```

Comment:
What about the `order id` or `swap id`. This needs to be part of the negotiation. 

If it is not, then during the execution, how can the maker tell the taker "param X did not match" if it does not know (for sure) what offer we are talking about?